diem/diem
# [x] integrate new test runner

TODO for this PR: 

- [x] Handle tests that leak child processes rather than hanging (deferred for now: https://github.com/diem/diem-devtools/issues/27)
- [x] doctests (deferred for now, we run unit tests and doctests in separate commands)
- [x] ignored tests
- [x] exit with non-zero code if any tests fail
- [x] Test against smoke tests
- [x] Fix issue with ignored tests in datatest-stable

---

# cargo nextest: a new way to run tests

Once this PR lands, Diem’s CI is going to start using `cargo nextest`, a brand-new test runner. This will speed up test runs in CI by around 60%: a full unit test run goes from around 24 minutes to 10 minutes.

## How nextest works

`cargo nextest` is a new way to run Rust tests. To understand what `nextest` does, it’s worth looking at how the current `cargo xtest` (and the upstream `cargo test`) works.

`cargo xtest` has three steps:

1. Build every specified test binary. (A test binary is an executable which contains tests in it.)
2. Run every test binary, one by one.
3. If a given test binary exits with a [zero (success) exit code](https://tldp.org/LDP/abs/html/exit-status.html), treat it as passed and move on to the next test binary, otherwise treat it as failed.

There are a couple of things to note here:

* `cargo xtest` treats every test binary as a black box — all it cares about are whether a binary exits with a zero (success) exit code or not. This provides maximum flexibility as to what a test binary does, but leaves performance on the table. [Constraints liberate, liberties constrain](https://tldp.org/LDP/abs/html/exit-status.html).
    * In practice, test executions are parallelized *within* a binary but not *across* binaries.
    * For example, if a test binary has 5 tests that take 5 seconds each, and one test that takes 120 seconds, execution of this test binary would take 120 seconds total and `cargo xtest` wouldn’t proceed to the next binary before it is done.
* Most of the output you see from `cargo xtest` is actually produced by the test binary itself, not by `cargo`. To get a better understanding of our tests, we need to do more than treat them as a black box: the output from individual binary invocations must be understood.

`cargo nextest` addresses these problems by internalizing the Rust test model more fully. It follows these steps:

1. Build every specified test binary (just like `cargo xtest`).
2. Run every test binary with `--list` and grab the list of tests from it.
3. In parallel (limited to the number of CPU cores) invoke every test in its own process, separately.

This lets us run tests across several binaries in parallel, and also provides nicer reporting (see screenshot). The speedup depends on the number of cores available. In our experiments in CI is around 60% — on my desktop it’s 75% or so.

![Screenshot of nextest output](https://user-images.githubusercontent.com/180618/112234435-c6998700-8bf9-11eb-8979-c4ce3d85a774.png)

As a result of this approach, we’ve already found bugs in two of Diem’s tests: [they were leaking processes](https://github.com/diem/diem/pull/7970) even on success. (Currently, those tests hang. I’d like to make it so that they fail with a “LEAK” status instead.)

## Future plans

`cargo nextest` is more than just a test runner: it is a *platform* to manage the health of our tests over time. Some upcoming work:

* The immediate next step is to add support for the [JUnit report format](https://llg.cubic.org/docs/junit/), a standard that is recognized by a wide variety of tools. This will let us off-the-shelf tools for performing analysis on our test results, and build our own tools wherever existing ones fall short. Expect every PR to have a test report associated with it by roughly the first week of April.
* The step after that would be to start aggregating test results over time, so we can track trends in our test performance and flakiness.
* `cargo nextest` currently doesn’t cover doctests or end-to-end tests — we run them in a separate `cargo xtest` invocation. Hope to add support for these sorts of tests soon.
* There are a number of further improvements we can prioritize according to our needs, such as:
    * configuring retries for tests
    * marking certain tests as heavier weight and consuming more “slots” in the test runner
    * choosing to run some tests in batches within the same binary execution, rather than invoking each test in its own binary
    * automatically marking and disabling flaky tests
    * recording tests that failed in the last run locally, and running them first

The current `cargo xtest` will continue to be available as is, but we’d like to make `cargo nextest` the best way to run tests locally and in CI. There’s a lot we can do here, and I’d love to hear your ideas!

## Acknowledgements and credits

`cargo nextest` is heavily inspired by Testpilot, Facebook’s internal test runner. It follows essentially the same approach as Testpilot. Thanks to all the wonderful people who have worked on it over the years.

Also thanks to Rex, Young and Sherry for their feedback and reviews.
